photon-client/src/integTest/scala/com/linkedin/photon/ml/DriverTest.scala

```diff
@@ -931,25 +963,131 @@ class DriverTest extends SparkTestUtils with TestTemplateWithTmpDir {
+        args += FieldNamesType.TRAINING_EXAMPLE.toString
+        args += CommonTestUtils.fromOptionNameToArg(OPTIMIZER_TYPE_OPTION)
+        args += OptimizerType.LBFGS.toString
+        args += CommonTestUtils.fromOptionNameToArg(TOLERANCE_OPTION)
+        args += 1e-12.toString
+        args += CommonTestUtils.fromOptionNameToArg(MAX_NUM_ITERATIONS_OPTION)
+        args += "1000"
+        args += CommonTestUtils.fromOptionNameToArg(REGULARIZATION_TYPE_OPTION)
+        args += RegularizationType.L2.toString
+        args += CommonTestUtils.fromOptionNameToArg(REGULARIZATION_WEIGHTS_OPTION)
+        args += lambdas.mkString(",")
+        args += CommonTestUtils.fromOptionNameToArg(USE_WARM_START)
+        args += useWarmStart
+
+        val completedJob = MockDriver.runLocally(
+          args.toArray,
+          sc,
+          expectedStages = stages,
+          expectedNumFeatures = LOGISTIC_EXPECTED_NUM_FEATURES,
+          expectedNumTrainingData = LOGISTIC_EXPECTED_NUM_TRAINING_DATA,
+          expectedIsSummarized = false,
+          expectedDiagnosticMode = DiagnosticMode.NONE)
+
+        val models: Array[(Double, GeneralizedLinearModel)] = loadAllModels(
+          new Path(outputDir, Driver.LEARNED_MODELS_TEXT).toString)
+        assertEquals(models.length, lambdas.length)
+        assertEquals(models.map(_._1), lambdas)
+
+        val metrics = completedJob.metrics
+        assertEquals(metrics.size, lambdas.length)
+        lambdas.foreach(l => assertTrue(metrics.contains(l)))
+
+        assertEquals(models.length, metrics.size)
+
+        (models, metrics)
+      }
+
+    val (coldModels, coldMetrics) = models(0)
+    val (warmModels, warmMetrics) = models(1)
+
+    // Check that the models have similar coefficients between warm-start and no-warm-start
+    // This check is within 1e-12 on all coefficients, which is stringent. It is possible to have
+    // same metrics (AUC...) with different models if features are correlated in the training data.
+    assertEquals(coldModels.length, warmModels.length)
+    coldModels.zip(warmModels).foreach { case ((coldLambda, coldModel), (warmLambda, warmModel)) =>
+      assertEquals(coldLambda, warmLambda)
+      VectorUtils.areAlmostEqual(coldModel.coefficients.means, warmModel.coefficients.means)
+    }
+
+    // Check that the models have identical metrics with similar values.
+    // Metrics should be equal in all cases, even if there are correlated features in the training data.
+    assertEquals(coldMetrics.size, warmMetrics.size)
+    coldMetrics.foreach { case (lambda, coldMetricsMap) =>
+      assertTrue(warmMetrics.contains(lambda))
+      warmMetrics(lambda).foreach { case (metricName, warmMetricValue) =>
+        assertTrue(coldMetricsMap.contains(metricName))
+        val coldMetricValue = coldMetricsMap(metricName)
+        val diff = (coldMetricValue - warmMetricValue) / math.min(coldMetricValue, warmMetricValue)
+        assertEquals(math.abs(diff), 0D, 1e-3)
```

Comment:
Not a var visible to the whole DriverTest: this epsilon is used only here. It had the name "WARM_START_EPSILON", which makes it clear it's specific to this case. Having the value here, since it is used only once and here, also has the merit of being immediately visible. Why introduce a variable that will need to be followed to see the value (for a value used ONCE and only HERE).